amaranth/vendor/lattice_nexus.py

```diff
@@ -0,0 +1,675 @@
+        """,
+        r"""
+        {{invoke_tool("programmer")}}
+            -oft -xcfsingle -revd -op "Fast Program"
+            -if {{name}}_impl/{{name}}_impl.bit -of {{name}}.xcf
+        """,
+    ]
+
+    # Common logic
+
+    def __init__(self, *, toolchain="Oxide"):
+        super().__init__()
+
+        assert toolchain in ("Oxide", "Radiant")
+        self.toolchain = toolchain
+
+    @property
+    def required_tools(self):
+        if self.toolchain == "Oxide":
+            return self._oxide_required_tools
+        if self.toolchain == "Radiant":
+            return self._radiant_required_tools
+        assert False
+
+    @property
+    def file_templates(self):
+        if self.toolchain == "Oxide":
+            return self._oxide_file_templates
+        if self.toolchain == "Radiant":
+            return self._radiant_file_templates
+        assert False
+
+    @property
+    def command_templates(self):
+        if self.toolchain == "Oxide":
+            return self._oxide_command_templates
+        if self.toolchain == "Radiant":
+            return self._radiant_command_templates
+        assert False
+
+    @property
+    def default_clk_constraint(self):
+        if self.default_clk == "OSC":
+            return Clock(150e6 / self.osc_div) # lscc/radiant/2022.1/docs/webhelp/eng/index.htm#page/Reference%20Guides/FPGA%20Libraries%20Reference%20Guide/Alphanumeric_Primitives_List.11.155.htm        
+        # For HFCLKOUT (created clock): ADJUSTED_PERIOD=6666.66667ps * HF_CLK_DIV
+        return super().default_clk_constraint
+
+    def create_missing_domain(self, name):
+        # Lattice ECP5 devices have two global set/reset signals: PUR, which is driven at startup
+        # by the configuration logic and unconditionally resets every storage element, and GSR,
+        # which is driven by user logic and each storage element may be configured as affected or
+        # unaffected by GSR. PUR is purely asynchronous, so even though it is a low-skew global
+        # network, its deassertion may violate a setup/hold constraint with relation to a user
+        # clock. To avoid this, a GSR/SGSR instance should be driven synchronized to user clock.
+        if name == "sync" and self.default_clk is not None:
+            m = Module()
+            if self.default_clk == "OSC":
+                if not hasattr(self, "osc_div"):
+                    raise ValueError("OSC divider (osc_div) must be an integer between 2 "
+                                     "and 128")
+                if not isinstance(self.osc_div, int) or self.osc_div < 2 or self.osc_div > 128:
+                    raise ValueError("OSC divider (osc_div) must be an integer between 2 "
+                                     "and 128, not {!r}"
+                                     .format(self.osc_div))
+                clk_i = Signal()
+                m.submodules += Instance("OSC", p_HF_CLK_DIV=self.osc_div, o_HFCLKOUT=clk_i)
```

Comment:
Have you checked the oscillator?

Reply:
with 
```
        platform.osc_div = 2
        platform.default_clk = "OSC"
```
in Blinky.py, seems I'm getting 
```
ERROR: Cell type 'OSC' instantiated as 'cd_sync.U$$0' is not supported by this device.
```
`top.debug.v` looks like:
```
  OSC #(
    .HF_CLK_DIV("2")
  ) \U$$0  (
    .HFCLKOUT(clk_i)
  );
 ```
 LMK if I'm missing something

Reply:
Let's just remove oscillator support then, it can be added later.

Reply:
Oops, I think I forgot to include the mapping from `OSC` to `OSC_CORE` in nextpnr

Reply:
I noticed LIteX uses OSCA --let me know if I should just use that